File: search/BinarySearch.java
import java.io.*;
import java.util.*;

class Solution {
    public static int binarySearchIterative(int what[], int find) {
        int startPoint = 0;
        int endPoint = what.length;
        int midPoint = (startPoint + endPoint) / 2;

        while (startPoint <= endPoint) {
            if (find > what[midPoint]) {
                startPoint = midPoint + 1;
            } else if (find < what[midPoint]) {
                endPoint = midPoint - 1;
            } else {
                return what[midPoint];
            }

            midPoint = (startPoint + endPoint) / 2;
        }

        return -1;
    }

    public static int binarySearch(int what[], int find, int startIndex, int endIndex) {
        int middleIndex = (startIndex + endIndex) / 2;

        if (startIndex <= endIndex) {
            if (find > what[middleIndex]) {
                return binarySearch(what, find, middleIndex + 1, endIndex);
            } else if (find < what[middleIndex]) {
                return binarySearch(what, find, startIndex, middleIndex - 1);
            } else {
                return what[middleIndex];
            }
        }

        return -1;
    }

    public static void main(String args[]) {
        int whatToSearch[] = { 1, 5, 7, 9, 11, 15, 56, 90 };
        System.out.println(binarySearch(whatToSearch, 90, 0, whatToSearch.length));
        System.out.println(binarySearchIterative(whatToSearch, 90));
    }
}

File: search/javascript/binary-search.js
function binarySearchIterative(whatToSearch, findValue) {
    let startIndex = 0;
    let endIndex = whatToSearch.length - 1;
    let middleIndex = Math.trunc((startIndex + endIndex) / 2);

    while (startIndex <= endIndex) {
        if (findValue < whatToSearch[middleIndex]) {
            endIndex = middleIndex - 1;
        } else if (findValue > whatToSearch[middleIndex]) {
            startIndex = middleIndex + 1;
        } else {
            return 1;
        }
        middleIndex = Math.trunc((startIndex + endIndex) / 2);
    }

    return -1;
}

function binarySearchRecursive(whatToSearch, findValue, startIndex, endIndex) {
    let middleIndex = Math.trunc((startIndex + endIndex) / 2);

    if (startIndex <= endIndex) {
        if (findValue < whatToSearch[middleIndex]) {
            return binarySearchRecursive(whatToSearch, findValue, startIndex, middleIndex - 1);
        } else if (findValue > whatToSearch[middleIndex]) {
            return binarySearchRecursive(whatToSearch, findValue, middleIndex + 1, endIndex);
        } else {
            return middleIndex;
        }
    }

    return -1;
}

console.log(`${binarySearchIterative([5, 6, 9, 11, 13], 14)}`);
console.log(`${binarySearchRecursive([1, 5, 6, 8, 10, 12, 15], 10, 0, 6)}`);
